src/core/providers/ogr/qgsogrproviderconnection.cpp

```diff
@@ -292,7 +293,33 @@ QgsVectorLayer *QgsOgrProviderConnection::createSqlVectorLayer( const QgsAbstrac
   QgsProviderMetadata *providerMetadata { QgsProviderRegistry::instance()->providerMetadata( QStringLiteral( "ogr" ) ) };
   Q_ASSERT( providerMetadata );
   QMap<QString, QVariant> decoded = providerMetadata->decodeUri( uri() );
-  decoded[ QStringLiteral( "subset" ) ] = sanitizeSqlForQueryLayer( options.sql ) ;
+
+  QString where;
+  QStringList columns;
+  QStringList tables;
+
+  QgsAbstractDatabaseProviderConnection::splitSimpleQuery( options.sql, columns, tables, where );
+
+  // We have two options here: if the original SQL is a plain SELECT * FROM table [WHERE ...] statement,
+  // we could turn this into a normal layer with a subset filter but this would be a bit inconsistent from a UX
+  // perspective because the SQL update menu would not be available, let's keep it a SQL layer always.
+
+  if ( !options.filter.isEmpty() )
+  {
+    if ( ! where.isEmpty() )
+    {
+      decoded[ QStringLiteral( "subset" ) ] = QStringLiteral( R"sql(%1 AND ( %2 ))sql" ).arg( sanitizeSqlForQueryLayer( options.sql ), options.filter );
```

Comment:
shouldn't that be ?
```suggestion
      decoded[ QStringLiteral( "subset" ) ] = QStringLiteral( R"sql(%1 AND ( %2 ))sql" ).arg( sanitizeSqlForQueryLayer( options.sql ), where );
```

and shouldn't %1 be surrounded by parenthesis ?